Zipper breakdown

ncreep opened the pull request:
An attempt to break `Zipper` into traits as per #76.
Not sure whether this is the most elegant way to do this...

As there's a bunch of ZipperXXX files now, I decided to move them into a separate folder.
Should we move them into a separate sub-package as well?

ncreep committed:
Introducing hidden elements into Zipper.

ncreep committed:
Adding a class for path transformation.

ncreep committed:
Fixing a spelling mistake.

ncreep committed:
Extracting node fetching into a separate object.

ncreep committed:
Adding Zipper shifting function.

ncreep committed:
Test for PathFetcher

ncreep committed:
Prettified tests for and fixed docs on PathTransformer.

ncreep committed:
Decoupling Zipper shifting function from PathTrasonfrmer.

ncreep committed:
Shifting preserves all times associated with holes.

ncreep committed:
Adding tests for Zipper hole shifting.

ncreep committed:
Whitespace fix.

ncreep committed:
Removing irrelevant type definition.

ncreep committed:
Adding a test to zipper.

ncreep committed:
Zipper shifting ignores empty paths.

ncreep committed:
Adding Zipper axes support.

ncreep committed:
Merge branch 'master' of git://github.com/djspiewak/anti-xml into axes

ncreep committed:
Moving zipper related classes into a separate folder.

ncreep committed:
Taking out Group methods overridden in Zipper into a separate trait.

ncreep committed:
Pulling Zipper hole shifting and axes into separate traits.

This removes the need in implicit conversion for Zipper axes support.

ncreep committed:
Adding package private modifier to all Zipper related traits.

ncreep committed:
Pulled Zipper unselection support into a separate trait.

ncreep committed:
Extracting zipper to hole map conversion methods into a trait.

ncreep committed:
Removing the `parent` method on Zipper, doesn't seem to be used
anywhere.

djspiewak commented:
Wow, I'm really on the ball here…  Taking a look now, should have some feedback tonight.

djspiewak committed:
Refactored all the zipper stuff into its own package

djspiewak committed:
Was over-zealous in my refactoring; ZipperSpecs should remain at root

djspiewak commented:
And just because I'm a horrible person, I didn't take care of this last week.  I have now though!  Everything looks good.  I do think you're right though that the zipper stuff (minus `Zipper.scala`) should be in its own package.  I went through and implemented that change, which is now pushed to my djspiewak:zipper-breakdown branch.  Let me know if there's anything I did wrong or something you want to change, otherwise I'll merge that branch into master this week.

Sorry again for taking so long!

ncreep commented:
I still have reservations about the readability of the whole thing, feels like the `Zipper` is being spread too thin between the numerous traits.
Also, before merging into master, the comments on #76 should be addressed (efficiency and usability).
I don't expect that the whole "zipper shifting" implementation is production ready, it was intended to be used only as a concept.

djspiewak commented:
> I still have reservations about the readability of the whole thing, feels like the Zipper is being spread too thin between the numerous traits.

That is the downside, to be sure.  It really only makes sense if we have something like `ZipperAxes` that we're trying to enable.  In other words, I wouldn't merge these changes separately; they belong together.

> Also, before merging into master, the comments on #76 should be addressed (efficiency and usability).

Addressing…

> I don't expect that the whole "zipper shifting" implementation is production ready, it was intended to be used only as a concept.

Gotcha.  How would you like to proceed?

ncreep commented:
I can think of multiple possible bottlenecks in the code:
- Freely hashing and sorting paths.
- Using `ZipperHoleMap` for depth first iteration. As noted in the code it was not optimized for it.
- Redundant traversals of the tree: first collecting paths, then converting to values.

Of course, I don't have any benchmarks to prove that, but I do know that addressing these issues would make the shifting code even uglier than it's already is. And as I suck at optimizing things, I would prefer someone else giving it a try.

The first step should probably be making some sort of pseudo benchmark, just to get a feel of whether the performance is anywhere near acceptable.

ncreep committed:
Whitespace fix.

ncreep committed:
Adding descendant Zipper axes.

Abusing PathCreator to reuse code, probably hurting performance.

ncreep committed:
Adding unselectAll to Zipper.

ncreep commented:
Adding the operations discussed in #76.

The same comments about performance still apply.
